code/messaging/src/phone/java/com/adobe/marketing/mobile/InAppNotificationHandler.java

```diff
@@ -9,98 +9,321 @@
+        }
+
+        final ArrayList<JsonUtilityService.JSONObject> ruleJsons = new ArrayList<>();
+        final ArrayList<Rule> parsedRules = new ArrayList<>();
+        // Loop through each rule in the payload and collect the rule json's present.
+        // Additionally, extract the image assets and build the asset list for asset caching.
+        for (final Map<String, Variant> currentItem : items) {
+            final Object dataObject = currentItem.get(MessagingConstants.EventDataKeys.Optimize.DATA);
+            final Map<String, String> data;
+            try {
+                if (dataObject instanceof Variant) {
+                    data = ((Variant) dataObject).getStringMap();
+                } else {
+                    data = (Map<String, String>) dataObject;
+                }
+            } catch (final VariantException e) {
+                Log.warning(LOG_TAG, "%s - Exception occurred when converting s VariantMap to StringMap: %s", SELF_TAG, e.getMessage());
                 return;
             }
-            final JSONArray items = payload.getJSONArray(MessagingConstants.EventDataKeys.Offers.ITEMS);
-            for (int index = 0; index < items.length(); index++) {
-                final String ruleJson = items.getJSONObject(index).getJSONObject(MessagingConstants.EventDataKeys.Offers.DATA).getString(MessagingConstants.EventDataKeys.Offers.CONTENT);
-                final JsonUtilityService.JSONObject rulesJsonObject = getJsonUtilityService().createJSONObject(ruleJson);
-                final Rule parsedRule = parseRuleFromJsonObject(rulesJsonObject);
-                if (parsedRule != null) {
-                    messagingModule.registerRule(parsedRule);
+            final String ruleJson = data.get(MessagingConstants.EventDataKeys.Optimize.CONTENT);
+            final JsonUtilityService.JSONObject ruleJsonObject = MessagingUtils.getJsonUtilityService().createJSONObject(ruleJson);
+            // we want to discard invalid jsons
+            if (ruleJsonObject != null) {
+                ruleJsons.add(ruleJsonObject);
+
+                // Parse the "img src" from each html payload in the current rule being processed then
+                // add the found assets to the imageAssetList so only current assets will be cached when the
+                // RemoteDownloader is used to download the image assets.
+                final String imageAsset = extractImageAssetFromJson(ruleJsonObject);
+                if (cacheUtilities.assetIsDownloadable(imageAsset)) {
+                    imageAssetList.add(imageAsset);
                 }
             }
-        } catch (JSONException e) {
-            Log.debug(LOG_TAG, "handleOfferNotification - JSON exception when attempting to retrieve rules json from the edge response event: %s", e.getLocalizedMessage());
+        }
+        // download and cache image assets
+        cacheUtilities.cacheImageAssets(imageAssetList);
```

Comment:
This function is doing lot of different things, let's break it down.
Two possible functions can be: `cacheImageAssets`, `registerRules`